src/main/java/io/orangebeard/listener/TestSuiteTree.java

```diff
@@ -0,0 +1,132 @@
+        this.testSuiteUUID = testSuiteUUID;
+    }
+
+    public String getName() { return testSuiteName; }
+    public String getNodeKey() { return nodeKey; }
+    public UUID getTestSuiteUUID() { return testSuiteUUID; }
+
+    @Override
+    public String toString() {
+        return String.format("(%s, %s, %s)", nodeKey, testSuiteName, testSuiteUUID);
+    }
+
+    /**
+     * Add a child node to the given tree node.
+     * The name of the child node must be unique among its siblings.
+     * In other words, a node should not have two or more children with the same value for "name".
+     * @param name Name of the new child node.
+     * @param nodeKey ID to register the test suite. It is the caller's responsibility that this is unique in the tree!
+     *                    It is advised to use <code>testSuite.UUID</code> for this ID.
+     * @param testSuiteUUID UUID of the test suite.
+     * @return An Optional containing the newly added node, or an empty Optional if there already was a child node with the given name.
+     */
+    public TestSuiteTree addChild(@NonNull String name, @NonNull String nodeKey, @NonNull UUID testSuiteUUID) {
+        // The field "name" should be unique among the children of a node.
+        if (getChildByName(name).isPresent()) {
+            return null;
+        }
+        // If there is not already a child node with the given name, create and add it.
+        TestSuiteTree child = new TestSuiteTree(name, nodeKey, testSuiteUUID);
+        children.add(child);
+        child.parent = this;
+        return child;
+    }
+
+    /** Find the subtree that was registered with the given ID.
+     * @param id Id of the subtree.
+     * @return The subtree with the given ID, or <code>null</code> if there is no such subtree.
+     */
+    public TestSuiteTree findSubtree(@NonNull String id) {
+        if (id.equals(nodeKey)) {
+            return this;
+        }
+
+        for (TestSuiteTree child: children) {
+            TestSuiteTree searchResult = child.findSubtree(id);
+            if (searchResult != null) {
+                return searchResult;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Finds the direct child node with the given name; or <code>null</code> if there is no such node.
+     * Names of child nodes should be unique among their siblings.
+     * @param name Name of the requested child node.
+     * @return An Optional containing the child node with the given name; or an empty Optional if there is no such node.
+     */
+    public Optional<TestSuiteTree> getChildByName(@NonNull String name) {
+        return children
+                .stream()
+                .filter(treeItem->name.equals(treeItem.getName()))
+                .findFirst()
+                ;
+    }
+
+    /**
+     * Make a list containing all the leaf nodes of the node.
+     * @return A list with all leaf nodes that descend from this node.
+     */
+    List<TestSuiteTree> getLeaves() {
```

Comment:
This method should be tested. And I have the feeling it's incomplete? Either you return the children or the leaf itself. Aren't you always missing the leaf itself when it has children?

Reply:
I've added a unit test. But the functionality is correct: leaf nodes are the nodes that have no children. So if the node has no children, it is itself a leaf. If it does have children, it recursively looks for the leaves in these subtrees instead.